pkg/server/server.go

```diff
@@ -0,0 +1,219 @@
+package server
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"github.com/armory/go-yaml-tools/pkg/secrets"
+	"io/ioutil"
+	"net/http"
+	"os"
+)
+
+type Server struct {
+	config *ServerConfig
+	server *http.Server
+}
+
+func NewServer(config *ServerConfig) *Server {
+	return &Server{
+		config: config,
+	}
+}
+
+// Start starts the server on the configured port
+func (s *Server) Start(router http.Handler) error {
+	if !s.config.Ssl.Enabled {
+		return s.startHttp(router)
+	}
+	return s.startTls(router)
+}
+
+func (s *Server) Shutdown(ctx context.Context) error {
+	return s.server.Shutdown(ctx)
+}
+
+func (s *Server) startHttp(router http.Handler) error {
+	s.server = &http.Server{
+		Addr:    s.config.GetAddr(),
+		Handler: router,
+	}
+	return s.server.ListenAndServe()
+}
+
+func (s *Server) startTls(router http.Handler) error {
+	tlsConfig, err := s.tlsConfig()
+	if err != nil {
+		return err
+	}
+
+	certMode := s.getClientCertMode()
+	if certMode != tls.NoClientCert {
+		// With mTLS, we'll parse our PEM to discover CAs with which to validate client certificates
+		caFile := s.config.Ssl.CAcertFile
+		if caFile == "" {
+			// Fall back to cert file - could be a combined PEM (e.g. self signed)
+			caFile = s.config.Ssl.CertFile
+		} else if err := checkFileReadable(caFile); err != nil {
+			return fmt.Errorf("error with certificate authority file %s: %w", caFile, err)
+		}
+
+		// Create a CA certificate pool and add our server certificate
+		caCert, err := ioutil.ReadFile(caFile)
+		if err != nil {
+			return err
+		}
+		caCertPool := x509.NewCertPool()
+		caCertPool.AppendCertsFromPEM(caCert)
+		tlsConfig.ClientCAs = caCertPool
+		tlsConfig.ClientAuth = certMode
+	}
+
+	// Discover the server name based on given certificates
+	tlsConfig.BuildNameToCertificate()
+
+	// Create a Server instance to listen on port 8443 with the TLS config
+	server := &http.Server{
+		Addr:      s.config.GetAddr(),
+		Handler:   router,
+		TLSConfig: tlsConfig,
+	}
+
+	// Listen to HTTPS connections with the server certificate and wait
+	return server.ListenAndServeTLS("", "")
```

Comment:
does this work? don't we care about offering up certs and verifying them?

Reply:
certs and keys are setup/decrypted and verified in `tlsConfig()`.